doc: how to create an ouput (#78)
AdRoll/baker: content/en/docs/How To/create_output.md

storing them, eventually sending the result (like a temporary file in the disk) to an Upload
component.

To create an output and make it available to Baker, one must:

* Implement the [Output](https://pkg.go.dev/github.com/AdRoll/baker#Output) interface
* Add an [`OutputDesc`](https://pkg.go.dev/github.com/AdRoll/baker#OutputDesc) for the output to
the available outputs in [`Components`](https://pkg.go.dev/github.com/AdRoll/baker#Components)

## The Output interface

```go
type Output interface {
	Run(in <-chan OutputRecord, upch chan<- string) error
	Stats() OutputStats
	CanShard() bool
}
```

The [Output interface](https://pkg.go.dev/github.com/AdRoll/baker#Output) must be implemented when
creating a new output component.

The `Run` function implements the component logic and gets a channel where it receives
[OutputRecord](https://pkg.go.dev/github.com/AdRoll/baker#OutputRecord) objects and a channel to
communicate to the Upload components what to upload.

`CanShard` is the function telling whether the output is able to manage sharding. Read the page
[dedicated to the sharding](/docs/how-to/sharding/) to go deeper in the topic.

> `Stats` is used to collect metrics, see the [dedicated page](/docs/how-to/metrics/).

Author: :+1:

## OutputDesc

```go
var MyOutputDesc = baker.OutputDesc{
	Name:   "MyOutput",
	New:    NewMyOutput,
    Config: &MyOutputConfig{},
    Raw:    true,
	Help:   "Help message for the output",
}
```

This object has a `Name`, that is used in the Baker configuration file to identify the output,
a constructor-like function (`New`), a config object (used to parse the output configuration in the
TOML file) and a help text that must help the users to use the component and its configuration
parameters. The `Raw` field tells whether the output expects to receive raw records in addition
to single fields (see below for details).

### Output constructor-like function

> The `New` key in the `OutputDesc` object represents the constructor-like function.

Author: ok, sounds right

The function receives an [OutputParams](https://pkg.go.dev/github.com/AdRoll/baker#OutputParams)
object and returns an instance of [Output](https://pkg.go.dev/github.com/AdRoll/baker#Output).

`OutputParams` contains the index of the output and the fields `FieldIndex` that the output will
receive during processing, in addition to the fields inherited from
[ComponentParams](https://pkg.go.dev/github.com/AdRoll/baker#ComponentParams).

`OutputParams.Index` indicates a unique index of the output process among the concurrent output
processes generated by Baker. The `procs` configuration can be used to tune the total number of
concurrent processes, see [Pipeline configuration](/docs/how-to/pipeline_configuration/) for details.  
Note that the output should extensively document in `OutputDesc.Help` if it is able to manage
concurrent processing or the user should configure it with a single process (`procs=1`).

> Read [Tuning concurrency](/docs/how-to/concurrency/) for an in-depth study on the subject.

Author: `guide to` added

`OutputParams.Fields` is a list of [FieldIndex](https://pkg.go.dev/github.com/AdRoll/baker#FieldIndex)
that the output will receive. Their position in this slice correspond to the fields in
[OutputRecord.Fields](https://pkg.go.dev/github.com/AdRoll/baker#OutputRecord), see below for
details.  
If, for any reason, the output needs to retrieve the fields name (like the SQLite output does to
get the columns names), then `OutputParams.FieldName` can be used.

### The output configuration and help

The output configuration object (`MyOutputConfig` in the previous example) must export all
configuration parameters that the user can set in the TOML topology file.

> Each key in the struct must include a `help` string tag and a `required` boolean tag, both are

Reviewer: key/field

`required` is not required though, it defaults to `false`.

Reviewer: there's also the default `tag` but as we said, we should just link to a section (that doesn't exist yet) which details configuration structs requirements and capabilities.

Author: fixed

mandatory.

> The former helps the user to understand the possible values of the field, the latter tells Baker
> whether to refuse a missing configuration param.

Reviewer: ```suggestion
All these parameters appear in the generated help. `help` should describe the parameter role and/or its possible values, `required` informs Baker it should refute configurations in which that field is not defined.
```

Author: done

## OutputRecord

The [OutputRecord](https://pkg.go.dev/github.com/AdRoll/baker#OutputRecord) channel received by
the output component is closed by Baker when all the records have been processed (in case of a batch

> execution) or when an interruption starts (in case of a SIGINT signal from the user).

Reviewer: I'm not sure we should write about interruption (since an output doesn't directly deal with that), neither about signal here (it' os-dependent).
IMHO when writing an output, an user shouldn't have to consider these, their horizon should stop at the limit of Output interface.
The only _contract_ the user should respect however is that they should return from `Run` as soon as possible once the channel has been closed, whatever the reason is (be it that there are no more records to process or the topology is terminating prematurely).

Author: ```The [OutputRecord](https://pkg.go.dev/github.com/AdRoll/baker#OutputRecord) channel received by
the output component can be closed by Baker at any time and the output should return from the `Run`
function as soon as possible when this happens.```

Reviewer: :+1:

> Until that moment, the output component must continuosly read new records, processing them.

Reviewer: ```suggestion
Until that moment, the output component must continuously read new records, processing them.
```

The `OutputRecord.Fields` slice contains the string values of the fields that the user choose to
send to the output configuring the
[`fields` key](http://localhost:1313/docs/how-to/pipeline_configuration/#components-configuration)
in the `[output]` section of the TOML topology file.

> The fields positions correspond to the `FieldIndex` received in `OutputParams.Fields` by the
> component constructor-like function.

Reviewer: ```suggestion
Fields are ordered in the same way than the slice of `FieldIndex` received in `OutputParams.Fields`.
```

> In case of a raw output, `OutputRecord.Record` contains the full record as a byte slice.  
> The presence of raw data doesn't change the fields field value, that depend on the output
> configuration in the TOML file.

Reviewer: Can we try put that in a single sentence for clarity?
```suggestion
In case of a raw output, `OutputRecord.Record` contains both the serialized record as a byte slice
and the field values. ```

Author: done

## Prepare data for uploading

The output component must always send the produced output to the upload component, that is optional

> but its presence is unknown to the output.

Reviewer: This is only true if the output produces files on the local filesystem, think about the dynamodb output for example.

Author: hell yeah! :laughing: fixing
